File: src/python/data_structures/binary_search_tree.py
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None
        self.size = 0
    
    def insert(self, data):
        self.root = self._insert_recursive(self.root, data)
        self.size += 1
    
    def _insert_recursive(self, root, data):
        if root is None:
            return TreeNode(data)
        
        if data < root.data:
            root.left = self._insert_recursive(root.left, data)
        elif data > root.data:
            root.right = self._insert_recursive(root.right, data)
        
        return root
    
    def search(self, data):
        return self._search_recursive(self.root, data) is not None
    
    def _search_recursive(self, root, data):
        if root is None or root.data == data:
            return root
        
        if data < root.data:
            return self._search_recursive(root.left, data)
        return self._search_recursive(root.right, data)
    
    def delete(self, data):
        self.root = self._delete_recursive(self.root, data)
        self.size -= 1
    
    def _delete_recursive(self, root, data):
        if root is None:
            return root
        
        if data < root.data:
            root.left = self._delete_recursive(root.left, data)
        elif data > root.data:
            root.right = self._delete_recursive(root.right, data)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            
            temp = self._find_min(root.right)
            root.data = temp.data
            root.right = self._delete_recursive(root.right, temp.data)
        
        return root
    
    def _find_min(self, root):
        while root.left is not None:
            root = root.left
        return root
    
    def inorder_traversal(self):
        result = []
        self._inorder_recursive(self.root, result)
        return result
    
    def _inorder_recursive(self, root, result):
        if root is not None:
            self._inorder_recursive(root.left, result)
            result.append(root.data)
            self._inorder_recursive(root.right, result)
    
    def preorder_traversal(self):
        result = []
        self._preorder_recursive(self.root, result)
        return result
    
    def _preorder_recursive(self, root, result):
        if root is not None:
            result.append(root.data)
            self._preorder_recursive(root.left, result)
            self._preorder_recursive(root.right, result)
    
    def postorder_traversal(self):
        result = []
        self._postorder_recursive(self.root, result)
        return result
    
    def _postorder_recursive(self, root, result):
        if root is not None:
            self._postorder_recursive(root.left, result)
            self._postorder_recursive(root.right, result)
            result.append(root.data)
    
    def __len__(self):
        return self.size
    
    def is_empty(self):
        return self.size == 0

File: src/java/data_structures/BinarySearchTree.java
package data_structures;

public class BinarySearchTree {
    private TreeNode root;
    private int size;
    
    private class TreeNode {
        int data;
        TreeNode left, right;
        
        TreeNode(int data) {
            this.data = data;
        }
    }
    
    public void insert(int data) {
        root = insertRecursive(root, data);
        size++;
    }
    
    private TreeNode insertRecursive(TreeNode root, int data) {
        if (root == null) return new TreeNode(data);
        
        if (data < root.data) {
            root.left = insertRecursive(root.left, data);
        } else if (data > root.data) {
            root.right = insertRecursive(root.right, data);
        }
        return root;
    }
    
    public boolean search(int data) {
        return searchRecursive(root, data) != null;
    }
    
    private TreeNode searchRecursive(TreeNode root, int data) {
        if (root == null || root.data == data) return root;
        
        if (data < root.data) return searchRecursive(root.left, data);
        return searchRecursive(root.right, data);
    }
    
    public void delete(int data) {
        root = deleteRecursive(root, data);
        size--;
    }
    
    private TreeNode deleteRecursive(TreeNode root, int data) {
        if (root == null) return root;
        
        if (data < root.data) {
            root.left = deleteRecursive(root.left, data);
        } else if (data > root.data) {
            root.right = deleteRecursive(root.right, data);
        } else {
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;
            
            TreeNode temp = findMin(root.right);
            root.data = temp.data;
            root.right = deleteRecursive(root.right, temp.data);
        }
        return root;
    }
    
    private TreeNode findMin(TreeNode root) {
        while (root.left != null) root = root.left;
        return root;
    }
    
    public void inorderTraversal() {
        inorderRecursive(root);
        System.out.println();
    }
    
    private void inorderRecursive(TreeNode root) {
        if (root != null) {
            inorderRecursive(root.left);
            System.out.print(root.data + " ");
            inorderRecursive(root.right);
        }
    }
    
    public void preorderTraversal() {
        preorderRecursive(root);
        System.out.println();
    }
    
    private void preorderRecursive(TreeNode root) {
        if (root != null) {
            System.out.print(root.data + " ");
            preorderRecursive(root.left);
            preorderRecursive(root.right);
        }
    }
    
    public void postorderTraversal() {
        postorderRecursive(root);
        System.out.println();
    }
    
    private void postorderRecursive(TreeNode root) {
        if (root != null) {
            postorderRecursive(root.left);
            postorderRecursive(root.right);
            System.out.print(root.data + " ");
        }
    }
    
    public int size() { return size; }
    public boolean isEmpty() { return size == 0; }
}